specs/IsNonClientRegionSupportEnabled.md

```diff
@@ -0,0 +1,176 @@
+# Background
+
+To improve the developer experience for customizing non-client regions, WebView2 is 
+working to support using DOM elements as non-client regions. We currently have limited 
+support for custom title bars aka draggable regions and are working on building out support 
+for caption controls and resize regions. 
+
+For security and flexibility, we want developers to be able to enable or disable all 
+custom non-client functionality per WebView. Non-client functionality will affect that 
+app window’s size and position so it’s important that developers can definitively 
+toggle access. This can be achieved in a limited way using a feature flag, but feature 
+flags are applied per WebView2 environment, thus, an API on the WebView2 to enable/disable 
+non-client support via a setting is the better solution.
+
+# Description
+`IsNonClientRegionSupportEnabled` defaults to `FALSE`. Disabling/Enabling 
+`IsNonClientRegionSupportEnabled` takes effect after the next navigation. Currently, draggable 
+regions is the only non-client experience we have implemented. Eventually, this setting will 
+expand to enable other non-client functionality, such as resize and caption controls. 
+
+When the setting is set to `TRUE`, then the following non-client region support for only the top 
+level document will be enabled:  
+* Web pages will be able to use the `app-region` CSS style. 
+* Draggable Regions declared with the CSS style `app-region: drag` will support title bar 
+functionality, for example, the dragging of the app window, the title bar context menu upon 
+right click, and the maximizing/restoring of the window size upon double click of the html 
+element. 
+
+When set to `FALSE`, then all non-client region support will be disabled.  
+* Web pages will not be able to use the `app-region` CSS style. 
+
+# Examples
+This example enables non-client region support for all pages on www.microsoft.com. 
+Pages on other origins will not have non-client region support enabled. 
+
+## Win32 C++
+```cpp 
+ScenarioNonClientRegionSupport::ScenarioNonClientRegionSupport(AppWindow* appWindow)
+    : m_appWindow(appWindow), m_webView(appWindow->GetWebView())
+{
+    
+    CHECK_FAILURE(m_webView->add_NavigationStarting(
```

Comment:
Text said that the property takes effect after the next navigation, but here we sneak it in just as navigation is starting.

When is the "next navigation" considered to have occurred? At the start of the Starting event? Or after all Starting events handlers have returned? (I'm sure this is doc'd somewhere since we have this "at the next navigation" policy in many places.)

Reply:
Currently we document the timing for updating settings [here](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2settings?view=webview2-1.0.2045.28#:~:text=Changes%20to%20IsGeneralAutofillEnabled%20and%20IsPasswordAutosaveEnabled%20apply%20immediately%2C%20while%20other%20setting%20changes%20made%20after%20NavigationStarting%20event%20do%20not%20apply%20until%20the%20next%20top%2Dlevel%20navigation): 

> Changes to IsGeneralAutofillEnabled and IsPasswordAutosaveEnabled apply immediately, while other setting changes made after NavigationStarting event do not apply until the next top-level navigation.

We may want to include this info in our navigation doc: https://learn.microsoft.com/en-us/microsoft-edge/webview2/concepts/navigation-events

Reply:
Please add a comment in the sample code where you set the setting that notes that the settings change takes effect after NavigationStarting for the next navigation.

Reply:
@david-risney We may need to update those common docs to clarify whether specifically it is

* changes made after the NavigationStarting event is raised
* changes made after all NavigationStarting event handlers have returned.

Basically, if I change something from *within* my NavigationStarting event handler, are those changes considered to have occurred before the navigation or after?

| Timeline | Remarks |
|-|-|
| Before NavigationStarting begins | In time |
| NavigationStarting begins (first handler is called) ||
| | If I change it here, is this in time? Or is it too late? |
| NavigationStarting ends (last handler returns) | |
| | Too late |

Wording could be

(If changes during the event handler are too late)
> Changes to `IsGeneralAutofillEnabled` and `IsPasswordAutosaveEnabled` apply immediately. Other changes to settings take effect when the NavigationStarting event is raised for the next top-level navigation.

(If changes during the event handler are in time)
> Changes to `IsGeneralAutofillEnabled` and `IsPasswordAutosaveEnabled` apply immediately. Other changes to settings take effect at the completion of the NavigationStarting event for the next top-level navigation.

Reply:
Based on how the sample behaves I believe that changes before and during the starting event are in time. We have a navigation completed event, I believe that this is the intended cutoff point, after which changes will not be applied.